File: riveredge-frontend/src/apps/kuaiplm/services/lab-judgment-rule.ts
/**
 * 实验判定规则 API（R-02）
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/lab-judgment-rules';

export interface LabJudgmentRule {
  id?: number;
  uuid?: string;
  rule_code?: string;
  rule_name?: string;
  version?: string;
  compare_type?: string;
  standard_min?: string | null;
  standard_max?: string | null;
  standard_value?: string | null;
  unit?: string | null;
  item_name?: string | null;
  is_active?: boolean;
  remarks?: string | null;
  created_by_name?: string | null;
  updated_by_name?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface LabJudgmentRuleOption {
  id: number;
  rule_code: string;
  rule_name: string;
  version: string;
  compare_type: string;
  standard_min?: string | null;
  standard_max?: string | null;
  standard_value?: string | null;
  unit?: string | null;
  item_name?: string | null;
  label: string;
}

function unwrapList(res: unknown): { items: LabJudgmentRule[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as LabJudgmentRule[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const labJudgmentRuleApi = {
  list: async (params?: Record<string, unknown>) => {
    const res = await api.get(BASE, { params });
    return unwrapList(res);
  },
  get: async (id: number) => (await api.get(`${BASE}/${id}`)) as LabJudgmentRule,
  create: async (data: Partial<LabJudgmentRule>) =>
    (await api.post(BASE, data)) as LabJudgmentRule,
  update: async (id: number, data: Partial<LabJudgmentRule>) =>
    (await api.put(`${BASE}/${id}`, data)) as LabJudgmentRule,
  revise: async (id: number, data?: Record<string, unknown>) =>
    (await api.post(`${BASE}/${id}/revise`, data || {})) as LabJudgmentRule,
  delete: async (id: number) => api.delete(`${BASE}/${id}`),
  options: async (params?: { keyword?: string; limit?: number }) => {
    const res = await api.get(`${BASE}/options`, { params });
    return (Array.isArray(res) ? res : []) as LabJudgmentRuleOption[];
  },
};

File: riveredge-backend/src/apps/kuaiplm/schemas/lab_judgment_rule.py
"""实验判定规则 Schema（R-02）。"""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class LabJudgmentRuleBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    rule_name: str = Field(..., max_length=200)
    version: str = Field("1", max_length=40)
    compare_type: str = Field("range", max_length=20)
    standard_min: Optional[str] = Field(None, max_length=80)
    standard_max: Optional[str] = Field(None, max_length=80)
    standard_value: Optional[str] = Field(None, max_length=80)
    unit: Optional[str] = Field(None, max_length=40)
    item_name: Optional[str] = Field(None, max_length=200)
    is_active: bool = True
    remarks: Optional[str] = None


class LabJudgmentRuleCreate(LabJudgmentRuleBase):
    rule_code: Optional[str] = Field(None, max_length=50)


class LabJudgmentRuleUpdate(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    rule_name: Optional[str] = Field(None, max_length=200)
    compare_type: Optional[str] = Field(None, max_length=20)
    standard_min: Optional[str] = Field(None, max_length=80)
    standard_max: Optional[str] = Field(None, max_length=80)
    standard_value: Optional[str] = Field(None, max_length=80)
    unit: Optional[str] = Field(None, max_length=40)
    item_name: Optional[str] = Field(None, max_length=200)
    is_active: Optional[bool] = None
    remarks: Optional[str] = None


class LabJudgmentRuleReviseRequest(BaseModel):
    """升版：复制当前规则为新版本（默认停用旧版）。"""

    version: Optional[str] = Field(None, max_length=40, description="新版本号；空则自增")
    deactivate_previous: bool = True
    rule_name: Optional[str] = Field(None, max_length=200)
    compare_type: Optional[str] = Field(None, max_length=20)
    standard_min: Optional[str] = Field(None, max_length=80)
    standard_max: Optional[str] = Field(None, max_length=80)
    standard_value: Optional[str] = Field(None, max_length=80)
    unit: Optional[str] = Field(None, max_length=40)
    item_name: Optional[str] = Field(None, max_length=200)
    remarks: Optional[str] = None


class LabJudgmentRuleResponse(LabJudgmentRuleBase):
    id: int
    uuid: str
    rule_code: str
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class LabJudgmentRuleListResponse(BaseModel):
    data: List[LabJudgmentRuleResponse] = Field(default_factory=list)
    total: int = 0
    success: bool = True


class LabJudgmentRuleOption(BaseModel):
    id: int
    rule_code: str
    rule_name: str
    version: str
    compare_type: str
    standard_min: Optional[str] = None
    standard_max: Optional[str] = None
    standard_value: Optional[str] = None
    unit: Optional[str] = None
    item_name: Optional[str] = None
    label: str
